event/src/unvalidated/builder.rs

```diff
@@ -203,9 +212,151 @@ impl<D> DataBuilder<DataBuilderWithData<D>> {
     }
 }
 
+/// Builder for constructing an [`unvalidated::TimeEvent`].
+#[derive(Default)]
+#[allow(private_bounds)]
+pub struct TimeBuilder<S: unvalidated::builder::TimeBuilderState> {
+    state: S,
+}
+
+/// State of the builder
+pub trait TimeBuilderState {}
+
+/// Initial state
+pub struct TimeBuilderEmpty;
+impl TimeBuilderState for TimeBuilderEmpty {}
+
+impl TimeBuilder<TimeBuilderEmpty> {
+    /// Specify the Cid of the init event for the stream.
+    pub fn with_id(self, id: Cid) -> TimeBuilder<TimeBuilderWithId> {
+        TimeBuilder {
+            state: TimeBuilderWithId { id },
+        }
+    }
+}
+
+/// State with id added
+pub struct TimeBuilderWithId {
+    id: Cid,
+}
+impl TimeBuilderState for TimeBuilderWithId {}
+
+impl TimeBuilder<TimeBuilderWithId> {
+    /// Specify details about the time transaction.
+    pub fn with_tx(
+        self,
+        chain_id: String,
+        tx_hash: Cid,
+        tx_type: String,
+    ) -> TimeBuilder<TimeBuilderWithTx> {
+        TimeBuilder {
+            state: TimeBuilderWithTx {
+                id: self.state.id,
+                chain_id,
+                tx_hash,
+                tx_type,
+            },
+        }
+    }
+}
+
+/// State with the transaction details.
+pub struct TimeBuilderWithTx {
+    id: Cid,
+    chain_id: String,
+    tx_hash: Cid,
+    tx_type: String,
+}
+impl TimeBuilderState for TimeBuilderWithTx {}
+
+impl TimeBuilder<TimeBuilderWithTx> {
+    /// Specify the root of the proof.
+    /// The index is an index into the edge that should be followed.
+    pub fn with_root(self, index: usize, edge: Ipld) -> TimeBuilder<TimeBuilderWithRoot> {
```

Comment:
the comments and names here confused me.  The thing called `edge` here isn't really the edge since it contains data not part of the proof path.  It really actually just is the IPLD encoding of the root node.  Same for the `with_edge` function below, the `edge` argument is really the entire IPLD object for this intermediate tree node.  Then the `index` is really the edge.  Can we rename `edge` to `node` and `index` to `edge_index`, or something like that?  Here and also for the `with_edge` function?

Reply:
This is good feedback, I wasn't really sure how to name these bits. I'll rename them.